R/s3utils.R

```diff
@@ -1,47 +1,66 @@
 REGION = "eu-west-1"
 
+# Checks if an expiry time has been reached, allowing for a window of five
+# minutes by default
+expired_auth <- function(expiry_t, window = 5 * 60) {
+  ifelse(
+    is.null(expiry_t),
+    TRUE,
+    as.POSIXct(Sys.time(), tz='UTC') + window > expiry_t
+  )
+}
+
 s3_svc <- function(region = REGION, ...) {
-  aws_role_arn <- Sys.getenv('AWS_ROLE_ARN')
+  # If the S3 service is yet to be set or the credentials are about to
+  # expire, get new credentials
+  if (is.null(Rs3tools.env$s3) |
+      (Rs3tools.env$temporary_authentication
+       & expired_auth(Rs3tools.env$authentication_expiry))) {
```

Comment:
I am just being picky,  if the only refreshing point is when the token is expired and if there is no way for the users using the package to reset it, then under some situations, the token somehow is broken, but the time is not expired? my point is that whether there is a way to force refreshing---> but as I said, it is only for discussion 😳

Reply:
Yes, that could lock someone out, I think a refresh credentials function should be possible.

Reply:
I think just adding a `force_cred_refresh` argument to the exported functions might be easier?  Or else this maybe highlights where the closure route is less flexible, because you can only alter the s3 variables from functions within `s3_gen(), whereas with your original method you'd just make a new function writing to the environment created in the package env...

Reply:
actually, I'm being silly - as long as there's a `force_cred_refresh` argument to the function that `s3_gen()` returns then you can just create a new function within which you call `svc_s3(force_cred_refresh = TRUE)` and then throw away the result (since you're just using it to refresh the s3 object in the enclosing environment).  I think that would work?

Reply:
That's exactly what I was thinking!

Reply:
Great - sorry for spelling it all out when you've already understood, mainly it's just useful for me to write it down to get a proper handle on all this.  Mostly I do modelling, so this stuff when developing a package as a service to (potentially) unknown users is outside my natural field!

Reply:
Added - https://github.com/moj-analytical-services/Rs3tools/pull/22